clients/src/main/java/org/apache/kafka/common/requests/AbstractRequest.java

```diff
@@ -25,40 +25,39 @@
 import java.nio.ByteBuffer;
 
 public abstract class AbstractRequest extends AbstractRequestResponse {
-    private final short version;
 
     public static abstract class Builder<T extends AbstractRequest> {
         private final ApiKeys apiKey;
-        private short version;
+        private final Short desiredVersion;
 
         public Builder(ApiKeys apiKey) {
+            this(apiKey, null);
+        }
+
+        public Builder(ApiKeys apiKey, Short desiredVersion) {
             this.apiKey = apiKey;
-            this.version = ProtoUtils.latestVersion(apiKey.id);
+            this.desiredVersion = desiredVersion;
         }
 
         public ApiKeys apiKey() {
             return apiKey;
         }
 
-        public Builder<T> setVersion(short version) {
-            this.version = version;
-            return this;
+        public short desiredOrLatestVersion() {
+            return desiredVersion == null ? ProtoUtils.latestVersion(apiKey.id) : desiredVersion;
```

Comment:
That's a good idea, here you go: https://github.com/apache/kafka/pull/2580/files